### PR TITLE
[WIP] More TS Binding Features

### PR DESCRIPTION
**Pull request by walnutwaldo**
Includes:
- Quantifiers w/o Patterns
- Improvements to the Arrays Typing
- Improvements to the FuncDecls Typing
- More high-level Z3 functions that exist in the Python API
- Small fixes to way C++ `uint32` is compiled to JS

This builds off of some previous work in adding Arrays. Some of those commits that messed up the formatting are unfortunately mixed in here still due to git issues. If we want a deep code dive and fixing the formatting on the main branch would be helpful for viewing diffs, we could run `npm run format` on the main branch code and then I can try to rebase/merge.

### CODE REVIEW

**Comment by walnutwaldo** [comment]
> @walnutwaldo please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="DeepReason Inc."]

**Comment by walnutwaldo** [comment]
@microsoft-github-policy-service agree company="DeepReason Inc."

**Comment by nunoplopes** [comment]
FWIW mixing whitespace changes with functional changes is a very good way to have your PR not reviewed by anyone. The size of the diff is a bit too large..

**Comment by walnutwaldo** [comment]
> FWIW mixing whitespace changes with functional changes is a very good way to have your PR not reviewed by anyone. The size of the diff is a bit too large..

Yeah, I hope to get the white space changes to disappear since those changes actually already exist on the main branch. I think it's an issue with the git histories not being matched though so I will need to figure out a good way to handle that.

File: src/api/js/src/high-level/types.ts [hunk 1]
```diff
@@ -1,59 +1,97 @@
...
-  ? IntNum<Name>
-  : S extends number | CoercibleRational
-  ? RatNum<Name>
-  : S extends boolean
-  ? Bool<Name>
-  : S extends Expr<Name>
-  ? S
-  : never;
+export type CoercibleToExprMap<S extends CoercibleToExpr<Name>, Name extends string = 'main'> =
+    S extends bigint
```
Comment (NikolajBjorner) [inline src/api/js/src/high-level/types.ts:54]: the style is inconsistent with the style used on line 39-50 and elsewhere. I guess an editor setting.

File: src/api/js/src/high-level/types.ts [hunk 2]
```diff
@@ -1044,73 +1616,92 @@ export interface AstMapCtor<Name extends string> {
...
-  values(): IterableIterator<Value>;
-  get(key: Key): Value | undefined;
-  set(key: Key, value: Value): void;
-  delete(key: Key): void;
-  clear(): void;
-  has(key: Key): boolean;
-  sexpr(): string;
+    extends Iterable<[Key, Value]> {
+    /** @hidden */
+    readonly __typename: 'AstMap';
```
Comment (NikolajBjorner) [inline src/api/js/src/high-level/types.ts:1621]: many line breaks are introduced (by the editor?) but not here.

**Comment by NikolajBjorner** [comment]
right, most style updates appear already in the master branch. Let them pass in before and didn't pay attention to the (large) change. There must be a ts style template that can be used as a convention for the code base.

**Comment by walnutwaldo** [comment]
> right, most style updates appear already in the master branch. Let them pass in before and didn't pay attention to the (large) change. There must be a ts style template that can be used as a convention for the code base.

Once I get the git histories to not be a pain to inspect, I can try to standardize the style convention. I believe `prettier` is supposed to be run here but it seems that for one reason or another `npm run format` isn't fixing inconsistent style.

**Comment by walnutwaldo** [comment]
Marking ready for review. Reiterating the top comment:

If we want a deep code dive and fixing the formatting on the main branch would be helpful for viewing diffs, we could run `npm run format` on the main branch code and then I can try to rebase/merge.

**Comment by NikolajBjorner** [comment]
I can merge this and then either of us can run npm run format and push updates?

**Comment by walnutwaldo** [comment]
> I can merge this and then either of us can run npm run format and push updates?

Sounds good. The current version should be formatted correctly I believe.

**Comment by NikolajBjorner** [comment]
C:\z3\src\api\js>npm run format

> z3-solver@0.1.0 format
> prettier --write '{./,src/,scripts/}**/*.{js,ts}'

[error] No files matching the pattern were found: "'{./,src/,scripts/}**/*.{js,ts}'".

**Comment by NikolajBjorner** [comment]
this one seems better

C:\z3\src\api\js\src>npm run format .

**Comment by NikolajBjorner** [comment]
it doesn't update anything, so the current files were already processed.

**Pull request merged**